### DISCUSSION TITLE
Change heading numbering within function?

### DISCUSSION THREAD

**Discussion by Holt59** [typst/typst#778] [Q&A] [answered]
I am writing a template with utility for other users and I'd like a function `appendices()` to start writing appendices with custom style, in particular, I'd like to 1) make a page break, 2) reset the heading counter and 3) change numbering to use `A` instead of `1`.

I currently have this:

```typ
#let appendices() = {
  pagebreak()
  counter(heading).update(0)
  counter("appendices").update(1)

  set heading(
    numbering: (..nums) => {
      let vals = nums.pos()
      let value = "ABCDEFGHIJ".at(vals.at(0) - 1)
      if vals.len() == 1 {
        return "APPENDIX " + value + ": "
      }
      else {
        return value + "." + nums.pos().slice(1).map(str).join(".")
      }
    }
  );
}
```

Part 1) and 2) work but the `set heading(...)` is not doing anything within the function. If I extract it and do a `#let` it works.

Any way to have it working within the function? Or something else? I tried doing a global numbering and use a counter to check if I was in appendices but it did not work.

**Comment by arj0019** [comment]
I want to have this same functionality as well. I will start with your code and see if I can figure out a solution.

**Reply by arj0019** [reply]
I was able to confirm that your numbering function works correctly.

**Reply by arj0019** [reply]
I tried many different fixes, but none worked. I think that there is an error in the compilation.

**Reply by arj0019** [reply]
I have submitted an issue for this discussion.

**Comment by PgBiel** [comment]
`set` and `show` are local to their current context, and thus are ineffective when the function ends. As suggested in https://github.com/typst/typst/issues/806#issuecomment-1509872977, you will have to accept the body to be modified by that `set` (to be part of the context) as a function parameter, and return it within a new content block, where the `set` rule will be applied. Something like

```rust
#let appendices(body) = {
  pagebreak()
  counter(heading).update(0)
  counter("appendices").update(1)

  set heading(
    numbering: (..nums) => {
      let vals = nums.pos()
      let value = "ABCDEFGHIJ".at(vals.at(0) - 1)
      if vals.len() == 1 {
        return "APPENDIX " + value + ": "
      }
      else {
        return value + "." + nums.pos().slice(1).map(str).join(".")
      }
    }
  );
  
  [#body]
}

#appendices[
   = First
   ... text ...
   = Second
   ... text ...
   = Third
   ... text ...
]
```

**Reply by Holt59** [reply]
Thanks for the clarification., That solution is far form ideal and I would barely usable... That's fine if you put three lines, but if you have multiple pages in appendices, that's not really clean.

It would be nice to have an option to do this more cleanly... I managed to do it by catching the counter value in the global `set heading` but that requires a custom `outline()` since you need the location to access the counter value.

**Reply by PgBiel** [reply, marked as answer]
> Thanks for the clarification., That solution is far form ideal and I would barely usable... 

Worry not - Typst has exactly what you need:

```rust
#let appendices(body) = {
  pagebreak()
  counter(heading).update(0)
  counter("appendices").update(1)

  set heading(
    numbering: (..nums) => {
      let vals = nums.pos()
      let value = "ABCDEFGHIJ".at(vals.at(0) - 1)
      if vals.len() == 1 {
        return "APPENDIX " + value + ": "
      }
      else {
        return value + "." + nums.pos().slice(1).map(str).join(".")
      }
    }
  );
  
  [#body]
}

#show: appendices

= First
... text ...
= Second
... text ...
= Third
... text ...
```

Using `#show: function` will wrap the whole document after it and give it as the first parameter to the function. So, the above will work, regardless of pages.

**Reply by Holt59** [reply]
Thanks, that's actually pretty nice. Any reason you are using `[#body]` instead of simply `body` there?

**Reply by PgBiel** [reply]
> Any reason you are using `[#body]` instead of simply `body` there?

Oh, nope, it works just fine as `body`.
